osu.Game/Database/RealmContextFactory.cs

```diff
@@ -0,0 +1,160 @@
+        private static readonly GlobalStatistic<int> contexts_created = GlobalStatistics.Get<int>("Realm", "Contexts (Created)");
+
+        /// <summary>
+        /// Get a context for the current thread for read-only usage.
+        /// If a <see cref="RealmWriteUsage"/> is in progress, the existing write-safe context will be returned.
+        /// </summary>
+        public Realm Get()
+        {
+            reads.Value++;
+            return getContextForCurrentThread();
+        }
+
+        /// <summary>
+        /// Request a context for write usage. Can be consumed in a nested fashion (and will return the same underlying context).
+        /// This method may block if a write is already active on a different thread.
+        /// </summary>
+        /// <returns>A usage containing a usable context.</returns>
+        public RealmWriteUsage GetForWrite()
+        {
+            writes.Value++;
+            Monitor.Enter(writeLock);
+            Realm context;
+
+            try
+            {
+                context = getContextForCurrentThread();
+
+                currentWriteTransaction ??= context.BeginWrite();
+            }
+            catch
+            {
+                // retrieval of a context could trigger a fatal error.
+                Monitor.Exit(writeLock);
+                throw;
+            }
+
+            Interlocked.Increment(ref currentWriteUsages);
+
+            return new RealmWriteUsage(context, usageCompleted) { IsTransactionLeader = currentWriteTransaction != null && currentWriteUsages == 1 };
+        }
+
+        private Realm getContextForCurrentThread()
+        {
+            var context = threadContexts.Value;
+
+            if (context?.IsClosed != false)
+                threadContexts.Value = context = createContext();
+
+            contexts_open.Value = threadContexts.Values.Count;
+
+            if (!refreshCompleted.Value)
+            {
+                // to keep things simple, realm refreshes are currently performed per thread context at the point of retrieval.
+                // in the future this should likely be run as part of the update loop for the main (update thread) context.
+                context.Refresh();
+                refreshCompleted.Value = true;
+            }
+
+            return context;
+        }
+
+        private Realm createContext()
+        {
+            contexts_created.Value++;
+
+            return Realm.GetInstance(new RealmConfiguration(storage.GetFullPath($"{database_name}.realm", true))
+            {
+                SchemaVersion = schema_version,
+                MigrationCallback = onMigration,
+            });
+        }
+
+        private void usageCompleted(RealmWriteUsage usage)
+        {
+            int usages = Interlocked.Decrement(ref currentWriteUsages);
+
+            try
+            {
+                rollbackRequired |= usage.RollbackRequired;
```

Comment:
this should probably have xmldoc that once this is set once, the entire transaction is dirtied and irreversibly marked for rollback, no matter the depth - can see myself getting potentially very confused by this (generally not how i expect nested transactions to work)

Reply:
an exception should be thrown outwards at this point. if you look how this is already used in BeatmapManager's import sequence, it should make that more clear. can agree that reading this locally does feel weird, but do note the exception flow.

again, this is from `DatabaseContextFactory` (not to say it's necessarily good code, but not new).